validation: remove BLOCK_FAILED_CHILD

stratospher opened the pull request:
Fixes https://github.com/bitcoin/bitcoin/issues/32173

even though we have a distinction between `BLOCK_FAILED_VALID` and `BLOCK_FAILED_CHILD` in the codebase,
we don't use it for anything. Whenever we check for BlockStatus, we use `BLOCK_FAILED_MASK` which encompasses both of them.

Since there is no functional difference between `BLOCK_FAILED_VALID` and `BLOCK_FAILED_CHILD` and it's added
code complexity to correctly categorise them (ex: https://github.com/bitcoin/bitcoin/pull/31405#discussion_r1914366243, https://github.com/bitcoin/bitcoin/pull/16856#issuecomment-565506585), we could just remove it.

Looking for conceptual feedback on whether it's better to improve handling of `BLOCK_FAILED_CHILD` in the codebase or remove `BLOCK_FAILED_CHILD`.

Of less relevance, but it would also fix a `reconsiderblock` crash that could happen in the situation mentioned in https://github.com/bitcoin/bitcoin/issues/32173#issuecomment-2767030982

Similar attempt in the past in https://github.com/bitcoin/bitcoin/pull/16856#issuecomment-568073859

stratospher committed:
validation: don't update BLOCK_FAILED_VALID to BLOCK_FAILED_CHILD in InvalidateBlock

- there is no functional difference between BLOCK_FAILED_VALID and BLOCK_FAILED_CHILD
and it's unnecessary code complexity to correctly categorise them.

stratospher committed:
refactor: use clearer variables in InvalidateBlock()

Improve upon the variable name for `invalid_walk_tip` to make the
InvalidateBlock logic easier to read. Block tip before disconnection
is now tracked directly via `disconnected_tip`, and `new_tip`
is the tip after the disconnect.

Co-authored-by: stickies-v <stickies-v@protonmail.com>

stratospher committed:
validation: stop using BLOCK_FAILED_CHILD

even though we have a distinction between BLOCK_FAILED_VALID
and BLOCK_FAILED_CHILD in the codebase, we don't use it for
anything. since there's no functional difference between them
and it's unnecessary code complexity to categorise them correctly,
just mark as BLOCK_FAILED_VALID instead.

stratospher committed:
validation: remove BLOCK_FAILED_MASK

since it's the same as BLOCK_FAILED_VALID now

stratospher committed:
validation: reset BLOCK_FAILED_CHILD to BLOCK_FAILED_VALID when loading from disk

- there maybe existing block indexes stored in disk with
  BLOCK_FAILED_CHILD
- since they don't exist anymore, clean up block index entries with
  BLOCK_FAILED_CHILD and reset it to BLOCK_FAILED_VALID.

Prabhat1308 reviewed (commented):
Concept ACK

I think removing is a better option. The only **weak** argument that we could have for not removing is that `BLOCK_FAILED_VALID` right now marks the start of the invalid block chain and is better for debugging manually. However , the burden of maintaining this distinction seems too high for debugging purposes with no actual benefit functionally.

mzumsande reviewed (commented):
Concept ACK

naiyoma reviewed (commented):
Concept ACK, 

Wondering if the CI failure is related to this issue. ->https://github.com/bitcoin/bitcoin/issues/32855

yuvicc reviewed (commented):
Concept ACK

The CI failure looks like due to #32855

stickies-v reviewed (commented):
Concept ACK for removing it.

naiyoma reviewed (commented):
TestedACK 06db700a1315bb655ac7fa12578c626990a22ea5


                      Block 2 (valid)
                        /         \
    (valid)Block height 3         Block height 3 (BLOCK_FAILED_CHILD)
                        |         |
    (valid) Block height 4        Block height 4(BLOCK_FAILED_CHILD)
                        |         |

Before removing BLOCK_FAILED_CHILD,this test(https://github.com/bitcoin/bitcoin/issues/32173#issue-2960274990) would fail at: `self.log.info(self.nodes[0].reconsiderblock(res["bestblock"]))`
(Reconsidering block at height 3 — which has BLOCK_FAILED_CHILD)
This failure happened because:

[This](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L5297) check fails, block 3 has BLOCK_FAILED_CHILD (not BLOCK_FAILED_VALID).

and so in this assertion -> https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L5363 

`pindex->nStatus & BLOCK_FAILED_MASK) != 0.`

after removing BLOCK_FAILED_CHILD:


 block 3 , status is  BLOCK_FAILED_VALID, and `pindexFirstInvalid` is set correctly.

mzumsande commented on src/validation.cpp:
```diff
@@ -3657,7 +3657,7 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
     if (pindex->nHeight == 0) return false;
 
-    CBlockIndex* to_mark_failed = pindex;
+    CBlockIndex* invalid_walk_tip = pindex;
```
commit message ca2421d52792b32754f1801492822b999839ec3a:
"since the previous commit removes BLOCK_FAILED_CHILD" - only the following commit removes it. Or do you mean "any special logic for BLOCK_FAILED_CHILD"?

mzumsande reviewed (commented):
Looks good - just some minor comments.

sipa reviewed (commented):
Concept ACK

stickies-v reviewed (commented):
Approach ACK 2e040d8b3c861c96c536754a82f4352b635c4d01

Intuitively it seems helpful to track something like `BLOCK_FAILED_CHILD`, but since it's been unused for so long it seems the use case isn't really there. Block validation logic is complex and critical and simplifying it is worth the effort, so I think this is the way to go.

stratospher committed:
test: check LoadBlockIndex correctly recomputes invalidity flags

Add a test for block index transitioning from legacy
BLOCK_FAILED_CHILD to BLOCK_FAILED_VALID behavior.

In the scenario where a valid block has a BLOCK_FAILED_CHILD
parent and a BLOCK_FAILED_VALID grandparent, ensure that all
three blocks are correctly marked as BLOCK_FAILED_VALID
after reloading the block index.

stickies-v commented on src/test/blockchain_tests.cpp:
```diff
@@ -145,6 +150,26 @@ BOOST_FIXTURE_TEST_CASE(invalidate_block, TestChain100Setup)
+        block_99->nStatus = (block_99->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;
+        block_100->nStatus = (block_100->nStatus & ~BLOCK_FAILED_VALID);
+
+        // Reload block index to recompute block status validity flags from disk.
```
Note that nothing is written to disk here, this is all happening in-memory. It doesn't look like the actual from-disk behaviour here is easy to test, because that's relying on `BlockManager` internals, so this is probably the closest we can test. (`nStatus` being a non-const public member while `m_dirty_blockindex` being private is... not great)

I think it makes sense to add this as a separate test case, it's not really related to `invalidateblock` and I find targeted tests easier to understand. Suggested diff with a few other improvements:

<details>
<summary>git diff on cfadc8038c</summary>

```diff
diff --git a/src/test/blockchain_tests.cpp b/src/test/blockchain_tests.cpp
index 5422095404..6b9fe75ea8 100644
--- a/src/test/blockchain_tests.cpp
+++ b/src/test/blockchain_tests.cpp
@@ -125,10 +125,6 @@ BOOST_FIXTURE_TEST_CASE(invalidate_block, TestChain100Setup)
     BlockValidationState state;
     auto* orig_tip = active.Tip();
 
-    CBlockIndex* block_100 = active[100];
-    CBlockIndex* block_99 = active[99];
-    CBlockIndex* block_98 = active[98];
-
     int height_to_invalidate = orig_tip->nHeight - 10;
     auto* tip_to_invalidate = active[height_to_invalidate];
     m_node.chainman->ActiveChainstate().InvalidateBlock(state, tip_to_invalidate);
@@ -150,26 +146,6 @@ BOOST_FIXTURE_TEST_CASE(invalidate_block, TestChain100Setup)
         WITH_LOCK(::cs_main, assert(pindex->nStatus & BLOCK_FAILED_VALID));
         pindex = pindex->pprev;
     }
-
-    {
-        // consider the chain of blocks block_98 <- block_99 <- block_100
-        // intentionally mark:
-        //   - block_98: BLOCK_FAILED_VALID (already marked as BLOCK_FAILED_VALID from previous test)
-        //   - block_99: BLOCK_FAILED_CHILD (clear BLOCK_FAILED_VALID from previous test and mark as BLOCK_FAILED_CHILD)
-        //   - block_100: not invalid (clear BLOCK_FAILED_VALID from previous test)
-        LOCK(::cs_main);
-        assert(block_98->nStatus & BLOCK_FAILED_VALID);
-        block_99->nStatus = (block_99->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;
-        block_100->nStatus = (block_100->nStatus & ~BLOCK_FAILED_VALID);
-
-        // Reload block index to recompute block status validity flags from disk.
-        m_node.chainman->LoadBlockIndex();
-
-        // check block_98, block_99, block_100 is marked as BLOCK_FAILED_VALID after reloading from disk
-        assert(block_98->nStatus & BLOCK_FAILED_VALID);
-        assert(block_99->nStatus & BLOCK_FAILED_VALID);
-        assert(block_100->nStatus & BLOCK_FAILED_VALID);
-    }
 }
 
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/test/validation_chainstatemanager_tests.cpp b/src/test/validation_chainstatemanager_tests.cpp
index bf440ca639..a5f4de17c6 100644
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -557,6 +557,28 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
     BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.size(), num_indexes - last_assumed_valid_idx + 1);
 }
 
+BOOST_FIXTURE_TEST_CASE(loadblockindex_invalid_descendants, TestChain100Setup)
+{
+    LOCK(Assert(m_node.chainman)->GetMutex());
+    // Consider the chain of blocks: grand_parent   <-      parent      <-       child
+    //                            BLOCK_FAILED_VALID   BLOCK_FAILED_CHILD     <not invalid>
+    // Test that when the block index is loaded, all blocks are marked as BLOCK_FAILED_VALID
+    auto* child{m_node.chainman->ActiveChain().Tip()};
+    auto* parent{child->pprev};
+    auto* grand_parent{parent->pprev};
+    grand_parent->nStatus = (grand_parent->nStatus | BLOCK_FAILED_VALID);
+    parent->nStatus = (parent->nStatus & ~BLOCK_FAILED_VALID) | BLOCK_FAILED_CHILD;
+    child->nStatus = (child->nStatus & ~BLOCK_FAILED_VALID);
+
+    // Reload block index to recompute block status validity flags.
+    m_node.chainman->LoadBlockIndex();
+
+    // check grand_parent, parent, child is marked as BLOCK_FAILED_VALID after reloading the block index
+    BOOST_CHECK(grand_parent->nStatus & BLOCK_FAILED_VALID);
+    BOOST_CHECK(parent->nStatus & BLOCK_FAILED_VALID);
+    BOOST_CHECK(child->nStatus & BLOCK_FAILED_VALID);
+}
+
 //! Ensure that snapshot chainstates initialize properly when found on disk.
 BOOST_FIXTURE_TEST_CASE(chainstatemanager_snapshot_init, SnapshotTestSetup)
 {

```
</details>

stringintech commented on src/validation.cpp:
```diff
@@ -3657,7 +3657,7 @@ bool Chainstate::InvalidateBlock(BlockValidationState& state, CBlockIndex* pinde
     if (pindex->nHeight == 0) return false;
 
-    CBlockIndex* to_mark_failed = pindex;
+    CBlockIndex* invalid_walk_tip = pindex;
```
> but the remaining code won't be called anyways

But previously we would call `InvalidChainFound()` on the last disconnected tip (and the rest of the code after the `m_chain.Contains(to_mark_failed)` check) even if we were interrupted; right?

stratospher commented on src/node/blockstorage.cpp:
```diff
@@ -463,7 +463,7 @@ bool BlockManager::LoadBlockIndex(const std::optional<uint256>& snapshot_blockha
             }
         }
-        if (!(pindex->nStatus & BLOCK_FAILED_MASK) && pindex->pprev && (pindex->pprev->nStatus & BLOCK_FAILED_MASK)) {
+        if (!(pindex->nStatus & BLOCK_FAILED_VALID) && pindex->pprev && (pindex->pprev->nStatus & BLOCK_FAILED_VALID)) {
```
@stringintech great catch! thank you! I've updated the PR to use the diff suggested above.

I forgot the 1 edge case scenario where this is possible -  https://github.com/bitcoin/bitcoin/issues/32173#issuecomment-2767030982.
I ran bitcoind with an inconsistent block index state from the functional test in the issue (`BLOCK_FAILED_CHILD` happens with valid parents) and current logic wouldn't work as you mentioned. but it works on the latest push now.

fanquake commented:
Plenty of conceptual buy-in here, just seems that review momentum has stalled. This fixes at least 1, if not two bugs, so would be good to land in `31.x`. I've added it to the milestone.

stickies-v commented on src/node/blockstorage.cpp:
```diff
@@ -487,10 +487,18 @@ bool BlockManager::LoadBlockIndex(const std::optional<uint256>& snapshot_blockha
+        }
+        if (!(pindex->nStatus & BLOCK_FAILED_VALID) && pindex->pprev && (pindex->pprev->nStatus & BLOCK_FAILED_VALID)) {
+            // All descendants of invalid blocks are invalid too.
+            pindex->nStatus = BLOCK_FAILED_VALID;
```
I don't think we mean to override the entire `nStatus` here, just bitflip? (sorry for getting this wrong in my [suggestion](https://github.com/bitcoin/bitcoin/pull/32950/changes#r2405502780) earlier)
```suggestion
            pindex->nStatus |= BLOCK_FAILED_VALID;
```

The below test change fails without this fix, and passes with.

<details>
<summary>git diff on 16636cc5b9</summary>

```diff
diff --git a/src/node/blockstorage.cpp b/src/node/blockstorage.cpp
index 45c72f5af8..218ee222ef 100644
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -495,7 +495,7 @@ bool BlockManager::LoadBlockIndex(const std::optional<uint256>& snapshot_blockha
         }
         if (!(pindex->nStatus & BLOCK_FAILED_VALID) && pindex->pprev && (pindex->pprev->nStatus & BLOCK_FAILED_VALID)) {
             // All descendants of invalid blocks are invalid too.
-            pindex->nStatus = BLOCK_FAILED_VALID;
+            pindex->nStatus |= BLOCK_FAILED_VALID;
             m_dirty_blockindex.insert(pindex);
         }
 
diff --git a/src/test/validation_chainstatemanager_tests.cpp b/src/test/validation_chainstatemanager_tests.cpp
index 40f99690ce..64aac67ccb 100644
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -601,6 +601,7 @@ BOOST_FIXTURE_TEST_CASE(loadblockindex_invalid_descendants, TestChain100Setup)
     //   - child: not invalid
     // Test that when the block index is loaded, all blocks are marked as BLOCK_FAILED_VALID
     auto* child{m_node.chainman->ActiveChain().Tip()};
+    BOOST_CHECK(child->nStatus & BLOCK_HAVE_DATA);
     auto* parent{child->pprev};
     auto* grand_parent{parent->pprev};
     grand_parent->nStatus = (grand_parent->nStatus | BLOCK_FAILED_VALID);
@@ -614,6 +615,8 @@ BOOST_FIXTURE_TEST_CASE(loadblockindex_invalid_descendants, TestChain100Setup)
     BOOST_CHECK(grand_parent->nStatus & BLOCK_FAILED_VALID);
     BOOST_CHECK(parent->nStatus & BLOCK_FAILED_VALID);
     BOOST_CHECK(child->nStatus & BLOCK_FAILED_VALID);
+
+    BOOST_CHECK(child->nStatus & BLOCK_HAVE_DATA);
 }
 
 //! Ensure that snapshot chainstate can be loaded when found on disk after a

```
</details>

stickies-v commented on src/node/blockstorage.cpp:
I think it's unsafe to place commit [validation: reset BLOCK_FAILED_CHILD to BLOCK_FAILED_VALID when loading from disk](https://github.com/bitcoin/bitcoin/pull/32950/changes/3a8226258e9a22ecad43d3e50daad8eb626d4e8b#r2814993781) after the [validation: remove BLOCK_FAILED_MASK](https://github.com/bitcoin/bitcoin/pull/32950/changes/42263bc30b8cb4655336fa991cf50c4975dcad73) commit. We should only remove `BLOCK_FAILED_MASK` when there can be no more instances of `BLOCK_FAILED_CHILD`. A good flow imo is: 
1) stop producing `BLOCK_FAILED_CHILD` flags
2) migrate legacy data
3) remove `BLOCK_FAILED_MASK`

stickies-v reviewed (approved):
ACK b9653739c403bbc9fc2dc932be053bcad7acbe89 modulo [updating the commit ordering](https://github.com/bitcoin/bitcoin/pull/32950#discussion_r2814993781).

Prior to this PR, code was setting `BLOCK_FAILED_CHILD`, but (outside of tests) never directly reading it - always through `BLOCK_FAILED_MASK`. This, in combination with multiple full re-reviews, gives me good confidence that dropping `BLOCK_FAILED_CHILD` is safe, even if the validation logic is complex and I don't always understand all the nuances.

mzumsande reviewed (commented):
ACK b9653739c403bbc9fc2dc932be053bcad7acbe89  - I agree that moving commit 3a8226258e9a22ecad43d3e50daad8eb626d4e8b to an earlier spot would be good though.

stratospher commented on src/validation.cpp:
```diff
@@ -3184,7 +3184,7 @@ CBlockIndex* Chainstate::FindMostWorkChain()
                 while (pindexTest != pindexFailed) {
                     if (fFailedChain) {
-                        pindexFailed->nStatus |= BLOCK_FAILED_CHILD;
+                        pindexFailed->nStatus |= BLOCK_FAILED_VALID;
```
nice find, I tried out a few scenarios like the one below and found that it doesn't need it since https://github.com/bitcoin/bitcoin/commit/f5149ddb9b7de3559943d7fda0f440e59413dfb5. 

- we receive headers A, B, C, D built on top of normal-chain -> A -> B -> C -> D
- we then receive blocks B, C, D
- we later receive A and find out it is invalid in `ConnectBlock`.

`ConnectBlock` calls `InvalidChainFound` and there's now a `SetBlockFailureFlags` inside since f5149ddb9b7de3559943d7fda0f440e59413dfb5 which marks B,C,D as invalid too. So we wouldn't enter a situation where `FindMostWorkChain` has to mark blocks as invalid since we already dealt with it when the invalid block was found. 

(need to think more about what happens during a race condition between `ForkMostWorkChain` and new header arriving - don't imagine it's a problem since we'd need `cs_main` anyways.)

will open a separate PR for this after some testing.

alexanderwiederin reviewed (commented):
Tested locally and reviewed all commits. Question on the migration code is inline.

ACK fb3e1bf9c9772631571ca46d29c50330ebf54dfd

alexanderwiederin commented on src/node/blockstorage.cpp:
```diff
@@ -487,10 +487,18 @@ bool BlockManager::LoadBlockIndex(const std::optional<uint256>& snapshot_blockha
+        if (pindex->nStatus & BLOCK_FAILED_CHILD) {
+            // BLOCK_FAILED_CHILD is deprecated, but may still exist on disk. Replace it with BLOCK_FAILED_VALID.
+            pindex->nStatus = (pindex->nStatus & ~BLOCK_FAILED_CHILD) | BLOCK_FAILED_VALID;
+            m_dirty_blockindex.insert(pindex);
```
Not sure how it's done here in core, but I would suggest we open a follow-up issue for the removal and reference it in the comment. What do you think?

mzumsande reviewed (commented):
re-ACK fb3e1bf9c9772631571ca46d29c50330ebf54dfd

sipa reviewed (commented):
Posthumous code review ACK fb3e1bf9c9772631571ca46d29c50330ebf54dfd

Did someone test that downgrading to previous releases works?
